### ISSUE TITLE
[C++20] [Modules] The macro of headers doesn't get processed as expected

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#56899]
According to http://eel.is/c++draft/cpp.import#5 and the example http://eel.is/c++draft/cpp.import#5, we could know that the point of undefinition should be exported too.

For example:
```C++
// foo.h
  #define D 45

  // bar.h
  #define D 45
  #undef D

  // use.cpp
  #include <iostream>
  import "foo.h"; // point of definition for `D`
  import "bar.h"; // point of undefinition for `D`

  int main() {
  #ifdef D
    std::cout << "Macro Value: " << D << "\n";
  #else
    std::cout << "Not defined.\n";
  #endif
  }
```

So the result should be "Not defined.\n". But the current result is "Macro Value: 45". 

The reproducing command is:
```
clang++ -std=c++20 -xc++-user-header --precompile foo.h -o foo.pcm
clang++ -std=c++20 -xc++-user-header --precompile bar.h -o bar.pcm
clang++ -std=c++20 -fmodule-file=foo.pcm -fmodule-file=bar.pcm use.cpp
```

**Comment by iains** [comment]
yes, the current implementation has a misunderstanding of the std.  I had already discussed this with Nathan - @ChuanqiXu9  unless you have time, I will take this one?

**Comment by cpplearner** [comment]
I think "Macro Value: 45" is the correct output, because the macro definition in `foo.h` is active at this point.

**Comment by ChuanqiXu9** [comment]
> 

Yeah, I am OK.

> I think "Macro Value: 45" is the correct output, because the macro definition in foo.h is active at this point.

But the point of undefinition is also active.

**Comment by cpplearner** [comment]
> > 
> 
> Yeah, I am OK.
> 
> > I think "Macro Value: 45" is the correct output, because the macro definition in foo.h is active at this point.
> 
> But the point of undefinition is also active.

But the point of undefinition is associated with a different macro definition. Note that [cpp.import]/5 says each `#define` introduces a distinct macro definition.

It would be different if bar.h imports foo.h

**Comment by ChuanqiXu9** [comment]
> > >

Oh, I think you are right. Do you mean the following example should return "Not defined.\n";.

```
  // foo.h
  #define D 45

  // bar.h
  import "foo.h";
  #undef D

  // use.cpp
  #include <iostream>
  import "foo.h"; // point of definition for `D`
  import "bar.h"; // point of undefinition for `D`

  int main() {
  #ifdef D
    std::cout << "Macro Value: " << D << "\n";
  #else
    std::cout << "Not defined.\n";
  #endif
  }
```

Then the current implementation looks good.

Then currently it looks the implementation of clang modules and header units are the same...

**Comment by iains** [comment]
> Then currently it looks the implementation of clang modules and header units are the same...

(unrelated to this PR, I think) 

Are you sure about that? - AFAIR, clang modules (and PCH) actually preserve the full history of the macro def/undefs in the PCH - rather than simply the final state.  I think that does make a difference to which identifiers are present after the header PCM is read.

----

I think there might be a bug in the implementation tho, when we go through a .ii file (i.e. -E and then -x...), because we will emit the macro def. but not the undef.  That is on my TODO to check.

**Comment by ChuanqiXu9** [comment]
> > Then currently it looks the implementation of clang modules and header units are the same...
> 
> (unrelated to this PR, I think)
> 
> Are you sure about that? - AFAIR, clang modules (and PCH) actually preserve the full history of the macro def/undefs in the PCH - rather than simply the final state. I think that does make a difference to which identifiers are present after the header PCM is read.

I'm not sure, really. I just failed to find the difference between preserving the full history and preserving the final state only...

> 
> I think there might be a bug in the implementation tho, when we go through a .ii file (i.e. -E and then -x...), because we will emit the macro def. but not the undef. That is on my TODO to check.

Do you mean if we get `use.ii` by `-E use.cpp`, the behavior of `use.ii` is different with `use.cpp`?
